docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,254 @@
+aforementioned ABCI++ methods on the `BaseApp` type. We describe the implementations
+of the two methods individually below.
+
+Prior to describing the implementation of the two new methods, it is important to
+note that the existing ABCI methods, `CheckTx`, `DeliverTx`, etc, still exist and
+serve the same functions as they do now.
+
+### `PrepareProposal`
+
+Prior to evaluating the decision for how to implement `PrepareProposal`, it is
+important to note that `CheckTx` will still be executed and will be responsible
+for evaluating transaction validity as it does now, with one very important
+_additive_ distinction.
+
+When executing transactions in `CheckTx`, the application will now add valid
+transactions, i.e. passing the AnteHandler, to its own mempool data structure.
+In order to provide a flexible approach to meet the varying needs of application
+developers, we will define both a mempool interface and a data structure utilizing
+Golang generics, allowing developers to focus only on transaction
+ordering. Developers requiring absolute full control can implement their own
+custom mempool implementation.
+
+> Note, there will most likely be a need to modify the AnteHandler method signature
+> to now accept a series of TxRecords.
+
+We define the general mempool interface as follows (subject to change):
+
+```go
+// MempoolTx we define an app-side mempool transaction interface that is as
+// minimal as possible, only requiring applications to define the size of the
+// transaction to be used when reaping and getting the transaction itself.
+// Interface type casting can be used in the actual app-side mempool implementation.
+type MempoolTx interface {
+	// Size returns the size of the transaction in bytes.
+	Size(codec.Codec) int
+	Tx() sdk.Tx
+}
+
+// PrepareTxRecord defines a wrapper around a MempoolTx that is returned from
+// PrepareProposal which includes an Action to inform Tendermint what to do with
+// the transaction.
+type PrepareTxRecord[T MempoolTx] struct {
+	Tx T
+	Action abci.TxAction
+}
+
+type Mempool[T MempoolTx] interface {
+	// Insert attempts to insert a MempoolTx into the app-side mempool returning
+	// an error upon failure.
+	Insert(sdk.Context, T) error
+	// ReapMaxBytes returns the next set of available transactions from the app-side
+	// mempool, up to maxBytes or until the mempool is empty. The application can
+	// decide to return transactions from its own mempool or from the incoming
+	// TxRecords or some combination of both. The notion of 'available' or 'next'
+	// is defined by the application's mempool implementation.
+	ReapMaxBytes(ctx sdk.Context, txRecords abci.TxRecords, maxBytes int) ([]PrepareTxRecord[T], error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+	// Remove attempts to remove a transaction from the mempool, returning an error
+	// upon failure.
+	Remove(sdk.Context, T) error
+}
```

Comment:
MempoolTx is a regular old interface, right? If so, why bring generics into the picture, versus e.g.

```go
type PrepareTxRecord struct {
	Tx     MempoolTx
	Action abci.TxAction
}

type Mempool interface {
	Insert(sdk.Context, MempoolTx) error
	ReapMaxBytes(ctx sdk.Context, txRecords abci.TxRecords, maxBytes int) ([]PrepareTxRecord, error)
	NumTxs() int
	Remove(sdk.Context, MempoolTx) error
}
```

?

Reply:
(This won't work directly, of course -- the Remove method implicitly requires that MempoolTx values are comparable, which isn't true.)

Reply:
I think I mentioned in a comment above from @williambanfield -- there really isn't a need for generics per se.